File: Android/app/src/main/java/com/example/app/views/java/pt/DimensBindingAdapters.java
/**
 * @author Bodenberg
 * GIT: https://github.com/www-virtues-ag/appdimens.git
 */
package com.example.app.views.java.pt;

import android.view.View;
import android.widget.TextView;
import androidx.databinding.BindingAdapter;

import ag.virtues.dimens.dynamic.code.VirtuesDimens;

/**
 * [EN] Custom Data Binding Adapters to apply dynamic dimensions from the VirtuesDimens library.
 *
 * [PT] Adaptadores de Data Binding personalizados para aplicar dimensões dinâmicas da biblioteca VirtuesDimens.
 */
public class DimensBindingAdapters {

    // [EN] --- Adapters for Layout Dimensions (Dp -> Px) ---
    // [PT] --- Adaptadores para Dimensões de Layout (Dp -> Px) ---

    /**
     * [EN] Sets the width of a View, converting the Dp Float value (e.g., 48f) to PX
     * using the dynamic adjustment from VirtuesDimensDynamic.
     * Usage in XML: app:dynamicWidthDp="@{48f}" or app:dynamicWidthDp="@{myFloatVariable}"
     *
     * [PT] Define a largura de uma View, convertendo o valor Dp Float (ex: 48f) para PX
     * usando o ajuste dinâmico do VirtuesDimensDynamic.
     * Uso no XML: app:dynamicWidthDp="@{48f}" ou app:dynamicWidthDp="@{minhaVariavelFloat}"
     */
    @BindingAdapter("app:dynamicWidthDp")
    public static void setDynamicWidth(View view, float dpValue) {
        // [EN] 1. Creates the adjustable Dp object (48.dp)
        //      2. Calls toPx(resources) to get the dynamically adjusted value in Pixels
        // [PT] 1. Cria o objeto Dp ajustável (48.dp)
        //      2. Chama toPx(resources) para obter o valor dinamicamente ajustado em Pixels
        float pxValue = VirtuesDimens.INSTANCE.dynamic(dpValue, false).toPx(view.getResources());

        view.getLayoutParams().width = (int) pxValue;
        view.requestLayout();
    }

    /**
     * [EN] Sets the height of a View, converting Dp Float to dynamic PX.
     *
     * [PT] Define a altura de uma View, convertendo Dp Float para PX dinâmico.
     */
    @BindingAdapter("app:dynamicHeightDp")
    public static void setDynamicHeight(View view, float dpValue) {
        float pxValue = VirtuesDimens.INSTANCE.dynamic(dpValue, false).toPx(view.getResources());

        view.getLayoutParams().height = (int) pxValue;
        view.requestLayout();
    }

    // [EN] --- Adapter for Text Size (Dp -> Sp/Px) ---
    // [PT] --- Adaptador para Tamanho de Texto (Dp -> Sp/Px) ---

    /**
     * [EN] Sets the text size (TextView), converting Dp Float to dynamic SP/PX.
     * VirtuesDimensDynamic.toSp() ensures that the scale adjustment is applied to the text.
     * Usage in XML: app:dynamicTextSizeDp="@{20f}"
     *
     * [PT] Define o tamanho do texto (TextView), convertendo Dp Float para SP/PX dinâmico.
     * O VirtuesDimensDynamic.toSp() garante que o ajuste de escala seja aplicado ao texto.
     * Uso no XML: app:dynamicTextSizeDp="@{20f}"
     */
    @BindingAdapter("app:dynamicTextSizeDp")
    public static void setDynamicTextSize(TextView textView, float dpValue) {
        // [EN] Converts Dynamic Dp to Scaleable Pixels (SP) in Pixels (Float)
        // [PT] Converte o Dp Dinâmico para Scaleable Pixels (SP) em Pixels (Float)
        float spValueInPx = VirtuesDimens.INSTANCE.dynamic(dpValue, false).toSp(textView.getResources());

        // [EN] Sets the text size using TypedValue.COMPLEX_UNIT_PX (raw pixels)
        // [PT] Define o texto usando TypedValue.COMPLEX_UNIT_PX (pixels brutos)
        textView.setTextSize(android.util.TypedValue.COMPLEX_UNIT_PX, spValueInPx);
    }
}

File: Android/app/src/main/java/ag/virtues/dimens/example/views/kotlin/en/DimensBindingAdapters.kt
/**
 * @author Bodenberg
 * GIT: https://github.com/www-virtues-ag/virtues-dimens.git
 */
package ag.virtues.dimens.example.views.kotlin.en

import android.view.View
import android.widget.TextView
import androidx.databinding.BindingAdapter
import ag.virtues.dimens.dynamic.code.Virtues

/**
 * [EN] Custom Data Binding Adapters to apply dynamic dimensions from the Virtues library.
 *
 * [PT] Adaptadores de Data Binding personalizados para aplicar dimensões dinâmicas da biblioteca Virtues.
 */
object DimensBindingAdapters {

    // [EN] --- Adapters for Layout Dimensions (Dp -> Px) ---
    // [PT] --- Adaptadores para Dimensões de Layout (Dp -> Px) ---

    /**
     * [EN] Sets the width of a View, converting the Float Dp value (e.g., 48f) to PX
     * using the dynamic adjustment from VirtuesDynamic.
     * XML usage: app:dynamicWidthDp="@{48f}" or app:dynamicWidthDp="@{myFloatVariable}"
     *
     * [PT] Define a largura de uma View, convertendo o valor Float Dp (ex: 48f) para PX
     * usando o ajuste dinâmico do VirtuesDynamic.
     * Uso no XML: app:dynamicWidthDp="@{48f}" ou app:dynamicWidthDp="@{myFloatVariable}"
     */
    @JvmStatic
    @BindingAdapter("app:dynamicWidthDp")
    fun setDynamicWidth(view: View, dpValue: Float) {
        // [EN] 1. Creates the adjustable Dp object (48.dp)
        //      2. Calls toPx(resources) to get the dynamically adjusted value in Pixels
        // [PT] 1. Cria o objeto Dp ajustável (48.dp)
        //      2. Chama toPx(resources) para obter o valor ajustado dinamicamente em Pixels
        val pxValue = Virtues.dynamic(dpValue).toPx(view.resources)

        view.layoutParams.width = pxValue.toInt()
        view.requestLayout()
    }

    /**
     * [EN] Sets the height of a View, converting Float Dp to dynamic PX.
     *
     * [PT] Define a altura de uma View, convertendo Float Dp para PX dinâmico.
     */
    @JvmStatic
    @BindingAdapter("app:dynamicHeightDp")
    fun setDynamicHeight(view: View, dpValue: Float) {
        val pxValue = Virtues.dynamic(dpValue).toPx(view.resources)

        view.layoutParams.height = pxValue.toInt()
        view.requestLayout()
    }

    // [EN] --- Adapter for Text Size (Dp -> Sp/Px) ---
    // [PT] --- Adaptador para Tamanho de Texto (Dp -> Sp/Px) ---

    /**
     * [EN] Sets the text size (TextView), converting Float Dp to dynamic SP/PX.
     * VirtuesDynamic.toSp() ensures that the scaling adjustment is applied to the text.
     * XML usage: app:dynamicTextSizeDp="@{20f}"
     *
     * [PT] Define o tamanho do texto (TextView), convertendo Float Dp para SP/PX dinâmico.
     * VirtuesDynamic.toSp() garante que o ajuste de escala seja aplicado ao texto.
     * Uso no XML: app:dynamicTextSizeDp="@{20f}"
     */
    @JvmStatic
    @BindingAdapter("app:dynamicTextSizeDp")
    fun setDynamicTextSize(textView: TextView, dpValue: Float) {
        // [EN] Converts Dynamic Dp to Scaleable Pixels (SP) in Pixels (Float)
        // [PT] Converte Dp Dinâmico para Pixels Escaláveis (SP) em Pixels (Float)
        val spValueInPx = Virtues.dynamic(dpValue).toSp(textView.resources)

        // [EN] Sets the text size using TypedValue.COMPLEX_UNIT_PX (raw pixels)
        // [PT] Define o tamanho do texto usando TypedValue.COMPLEX_UNIT_PX (pixels brutos)
        textView.setTextSize(android.util.TypedValue.COMPLEX_UNIT_PX, spValueInPx)
    }
}